src/dir/glob.cr

```diff
@@ -1,266 +1,297 @@
 class Dir
+  # Returns an array of all files that match against any of *patterns*.
+  #
+  # The pattern syntax is similar to shell filename globbing, see `File.match?` for details.
+  #
+  # NOTE: Path separator in patterns needs to be always `/`. The returned file names use system-specific path separators.
   def self.[](*patterns) : Array(String)
     glob(patterns)
   end
 
+  # ditto
   def self.[](patterns : Enumerable(String)) : Array(String)
     glob(patterns)
   end
 
-  def self.glob(*patterns) : Array(String)
-    glob(patterns)
+  # Returns an array of all files that match against any of *patterns*.
+  #
+  # The pattern syntax is similar to shell filename globbing, see `File.match?` for details.
+  #
+  # If *match_hidden* is `true` the pattern will match hidden files and folders.
+  #
+  # NOTE: Path separator in patterns needs to be always `/`. The returned file names use system-specific path separators.
+  def self.glob(*patterns, match_hidden = false) : Array(String)
+    glob(patterns, match_hidden: match_hidden)
   end
 
-  def self.glob(*patterns)
-    glob(patterns) do |pattern|
-      yield pattern
-    end
-  end
-
-  def self.glob(patterns : Enumerable(String)) : Array(String)
+  # ditto
+  def self.glob(patterns : Enumerable(String), match_hidden = false) : Array(String)
     paths = [] of String
-    glob(patterns) do |path|
+    glob(patterns, match_hidden: match_hidden) do |path|
       paths << path
     end
     paths
   end
 
-  def self.glob(patterns : Enumerable(String))
-    special = {'*', '?', '{', '}'}
-    cwd = self.current
-    root = File::SEPARATOR_STRING
-    patterns.each do |ptrn|
-      next if ptrn.empty?
-
-      starts_with_path_separator = ptrn.starts_with?(File::SEPARATOR)
-
-      # If the pattern ends with a file separator we only want leaf
-      # directories: regex is the same as the one without that last char
-      wants_dir = false
-      if ptrn.ends_with?(File::SEPARATOR)
-        ptrn = ptrn.rchop
-        wants_dir = true
-      end
+  # Yields all files that match against any of *patterns*.
+  #
+  # The pattern syntax is similar to shell filename globbing, see `File.match?` for details.
+  #
+  # If *match_hidden* is `true` the pattern will match hidden files and folders.
+  #
+  # NOTE: Path separator in patterns needs to be always `/`. The returned file names use system-specific path separators.
+  def self.glob(*patterns, match_hidden = false, &block : String -> _)
+    glob(patterns, match_hidden: match_hidden) do |path|
+      yield path
+    end
+  end
 
-      recursion_depth = ptrn.count(File::SEPARATOR)
+  # ditto
+  def self.glob(patterns : Enumerable(String), match_hidden = false, &block : String -> _)
+    Globber.new(match_hidden: match_hidden).glob(patterns) do |path|
```

Comment:
What's the benefit of this object vs just passing options to a `glob` method?

Reply:
True. I think I was expecting this to be more complicated - or it even might have been so at the beginning, not sure.